In [1]:
import pandas as pd
import Funciones as F

In [2]:
customers = pd.read_excel('bc_customers_dataset.xlsx')
customers

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  
0                         14409                 franca             SP  
1                          9790  sao bernardo do campo             SP  
2                          1151              sao paulo             SP  
3                          8775        mogi das cruzes             SP  
4                         13056               campinas             SP  
...                         ...                    ...            ...  
99436                      3937              sao paulo             SP  
99437                      6764        taboao da serra             SP  
99438                     60115              fortaleza             CE  
99439                     92120                 canoas             RS  
99440                      6703                  cotia             SP  

[99441 rows x 5 columns]

In [3]:
orders = pd.read_excel('bc_orders_dataset.xlsx')
orders

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99276  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99277  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99278  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99279  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99280  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status   order_approved_at order_received_by_local_operator_date  \
0        delivered 2017-10-02 11:07:15                   2017-10-04 19:55:00   
1        delivered 2018-07-26 03:24:27                   2018-07-26 14:31:00   
2        delivered 2018-08-08 08:55:23                   2018-08-08 13:50:00   
3        delivered 2017-11-18 19:45:59                   2017-11-22 13:39:59   
4        delivered 2018-02-13 22:20:29                   2018-02-14 19:46:34   
...            ...                 ...                                   ...   
99276    delivered 2017-03-09 09:54:05                   2017-03-10 11:18:03   
99277    delivered 2018-02-06 13:10:37                   2018-02-07 23:22:42   
99278    delivered 2017-08-27 15:04:16                   2017-08-28 20:52:26   
99279    delivered 2018-01-08 21:36:21                   2018-01-12 15:35:03   
99280    delivered 2018-03-09 11:20:28                   2018-03-09 22:11:59   

      order_delivered_to_customer_date order_delivery_promise_date  
0                  2017-10-10 21:25:13                  2017-10-18  
1                  2018-08-07 15:27:45                  2018-08-13  
2                  2018-08-17 18:06:29                  2018-09-04  
3                  2017-12-02 00:28:42                  2017-12-15  
4                  2018-02-16 18:17:02                  2018-02-26  
...                                ...                         ...  
99276              2017-03-17 15:08:01                  2017-03-28  
99277              2018-02-28 17:37:56                  2018-03-02  
99278              2017-09-21 11:24:17                  2017-09-27  
99279              2018-01-25 23:32:54                  2018-02-15  
99280              2018-03-16 13:08:30                  2018-04-03  

[99281 rows x 7 columns]

-------------

In [4]:
F.analizar_datos(customers)

Nombre Tipos de Datos Únicos  % de Valores No Nulos  \
0               customer_id       [<class 'str'>]                  100.0   
1        customer_unique_id       [<class 'str'>]                  100.0   
2  customer_zip_code_prefix       [<class 'int'>]                  100.0   
3             customer_city       [<class 'str'>]                  100.0   
4            customer_state       [<class 'str'>]                  100.0   

   % de Valores Nulos  Cantidad de Valores Nulos  
0                 0.0                          0  
1                 0.0                          0  
2                 0.0                          0  
3                 0.0                          0  
4                 0.0                          0

In [5]:
F.analizar_datos(orders)

Nombre  \
0                               order_id   
1                            customer_id   
2                           order_status   
3                      order_approved_at   
4  order_received_by_local_operator_date   
5       order_delivered_to_customer_date   
6            order_delivery_promise_date   

                               Tipos de Datos Únicos  % de Valores No Nulos  \
0                                    [<class 'str'>]                 100.00   
1                                    [<class 'str'>]                 100.00   
2                                    [<class 'str'>]                 100.00   
3  [<class 'pandas._libs.tslibs.timestamps.Timest...                 100.00   
4  [<class 'pandas._libs.tslibs.timestamps.Timest...                  98.35   
5  [<class 'pandas._libs.tslibs.timestamps.Timest...                  97.16   
6  [<class 'pandas._libs.tslibs.timestamps.Timest...                 100.00   

   % de Valores Nulos  Cantidad de Valores Nulos  
0                0.00                          0  
1                0.00                          0  
2                0.00                          0  
3                0.00                          0  
4                1.65                       1637  
5                2.84                       2819  
6                0.00                          0

------

### ``Customer_id``

In [6]:
customers['customer_id'].duplicated().value_counts()

customer_id
False    99441
Name: count, dtype: int64

### ``Customer_unique_id``

In [7]:
customers['customer_unique_id'].duplicated().value_counts()

customer_unique_id
False    96096
True      3345
Name: count, dtype: int64

Calculamos cuanto representan estos duplicados del total

In [8]:
Porcentaje_unique = (3345*100/len(customers))
Porcentaje_unique

3.3638036624732255

In [9]:
duplicados_unique = customers[customers.duplicated(subset=['customer_unique_id'], keep=False)]
duplicados_unique_ordenados = duplicados_unique.sort_values(by='customer_unique_id')
duplicados_unique_ordenados

customer_id                customer_unique_id  \
35608  24b0e2bd287e47d54d193e7bbb51103f  00172711b30d52eea8b313a7f2cced02   
19299  1afe8a9c67eec3516c09a8bdcc539090  00172711b30d52eea8b313a7f2cced02   
20023  1b4a75b3478138e99902678254b260f4  004288347e5e88a27ded2bb23747066c   
22066  f6efe5d5c7b85e12355f9d5c3db46da2  004288347e5e88a27ded2bb23747066c   
72451  49cf243e0d353cd418ca77868e24a670  004b45ec5c64187465168251cd1c9c2f   
...                                 ...                               ...   
75057  1ae563fdfa500d150be6578066d83998  ff922bdd6bafcdf99cb90d7f39cea5b3   
27992  bec0bf00ac5bee64ce8ef5283051a70c  ff922bdd6bafcdf99cb90d7f39cea5b3   
79859  d064be88116eb8b958727aec4cf56a59  ff922bdd6bafcdf99cb90d7f39cea5b3   
64323  4b231c90751c27521f7ee27ed2dc3b8f  ffe254cc039740e17dd15a5305035928   
12133  0088395699ea0fcd459bfbef084997db  ffe254cc039740e17dd15a5305035928   

       customer_zip_code_prefix customer_city customer_state  
35608                     45200        jequie             BA  
19299                     45200        jequie             BA  
20023                     26220   nova iguacu             RJ  
22066                     26220   nova iguacu             RJ  
72451                     57055        maceio             AL  
...                         ...           ...            ...  
75057                     17340  barra bonita             SP  
27992                     17340  barra bonita             SP  
79859                     17340  barra bonita             SP  
64323                     37640       extrema             MG  
12133                     37640       extrema             MG  

[6342 rows x 5 columns]

Después de un análisis exhaustivo de los datos, se ha determinado que un 3,36% de los registros presentan duplicados en esta columna. Al ser un porcentaje bajo del total he tomado la decisión de eliminarlos debido a que la presencia de duplicados puede generar inconsistencias en el análisis y afectar la precisión de los resultados. Eliminar estos duplicados contribuirá a mantener la coherencia y la fiabilidad de los datos.

In [10]:
customers = customers.drop_duplicates(subset='customer_unique_id')

Revisamos la longitud del Dataset nuevamente para corroborar que se eliminaron los duplicados.

In [11]:
len(customers)

96096

### ``customer_zip_code_prefix``

In [12]:
customers['customer_id'].duplicated().value_counts()

customer_id
False    96096
Name: count, dtype: int64

### ``customer_city``

Elegiremos los primero 10 valores, ya que contamos con 4118 ciudades que se repiten, en lo cual no tiene una representacion que sea de valor significativo en porcentaje como las primeras 10 ciudades.

In [20]:
customers['customer_city'].value_counts()

customer_city
sao paulo                  14971
rio de janeiro              6611
belo horizonte              2671
brasilia                    2066
curitiba                    1462
                           ...  
pontes gestal                  1
nossa senhora das dores        1
francisco santos               1
engenheiro balduino            1
eugenio de castro              1
Name: count, Length: 4118, dtype: int64

In [17]:
top_10_ciudades = customers['customer_city'].value_counts().head(10)
top_10_ciudades

customer_city
sao paulo                14971
rio de janeiro            6611
belo horizonte            2671
brasilia                  2066
curitiba                  1462
campinas                  1397
porto alegre              1326
salvador                  1208
guarulhos                 1151
sao bernardo do campo      908
Name: count, dtype: int64

### ``customer_state``

In [22]:
customers['customer_state'].duplicated().value_counts()

customer_state
True     96069
False       27
Name: count, dtype: int64

In [23]:
customers['customer_state'].value_counts()

customer_state
SP    40295
RJ    12377
MG    11255
RS     5277
PR     4882
SC     3529
BA     3276
DF     2073
ES     1963
GO     1951
PE     1604
CE     1311
PA      949
MT      875
MA      725
MS      693
PB      519
PI      482
RN      474
AL      401
SE      341
TO      273
RO      239
AM      143
AC       77
AP       67
RR       45
Name: count, dtype: int64

A modo de verificar la consistencia de los datos, se le pregunto a chatgpt por la cantidad de ciudades y estados que posee Brasil, en el mismo corroboramos que efectivamente Brasil tiene 27 estados incluyendo Brasilia que es un Distrito Federal y tiene 5570 municipios los cuales estarian haciendo referencia como customer_city en nuestro dataset. Vemos que los duplicados en este caso estan siendo correctos los valores, otorgando un numero por debajo de 4118 municipios y 27 estados.

Conclusion:

- No se les llama ciudades sino Municipios, corregiremos el nombre de la columna.
- Se tienen clientes de todas los municipios de Brasil.
- El municipio que mas clientes tiene es Sao Paulo con un 15% del total.
- A su vez Sao Paulo tambien es el que mas clientes posee con un 42% del total de clientes.

FUENTE: https://chatgpt.com/share/9822ce44-a0e4-454f-9659-64eec23b63aa 